### PR TITLE
ENH: more bti tweaks

### PR DESCRIPTION
**Pull request by dengemann**
to allow not converting names + index order for private functions.

cc @Eric89GXL @agramfort

### CODE REVIEW

File: mne/io/bti/bti.py [hunk 1]
```diff
@@ -1009,204 +1015,16 @@ def __init__(self, pdf_fname, config_fname='config',
...
+        info, bti_info = _get_bti_info(
+            pdf_fname=pdf_fname, config_fname=config_fname,
+            head_shape_fname=head_shape_fname, rotation_x=rotation_x,
+            translation=translation, convert=convert, ecg_ch=ecg_ch,
+            eog_ch=eog_ch)
         logger.info('Reading raw data from %s...' % pdf_fname)
         data = _read_data(bti_info)
         assert len(data) == len(info['ch_names'])
         self._projector_hashes = [None]
+        self.bti_ch_labels = [c['chan_label'] for c in bti_info['chs']]
```
Comment (Eric89GXL) [inline mne/io/bti/bti.py]: this should probably be private, no?

File: mne/io/bti/bti.py [hunk 2]
```diff
@@ -1216,6 +1034,209 @@ def by_name(x):
...
+                       'ables is not supported. For critical use cases '
+                       '\nplease take into account the MNE command '
+                       '\'mne_create_comp_data\' to include weights as '
+                       'printed out \nby the 4D \'print_table\' routine.')
+
+    # check that the info is complete
+    assert set(RAW_INFO_FIELDS) == set(info.keys())
+
+    # check nchan is correct
+    assert len(info['ch_names']) == info['nchan']
```
Comment (Eric89GXL) [inline mne/io/bti/bti.py]: better just to do `info._check_consistency()` instead

**Comment by Eric89GXL** [comment]
Otherwise LGTM

File: mne/io/bti/bti.py (hunk 1 above)
Comment (dengemann) [inline mne/io/bti/bti.py]: it has been exisiting already like this, only moving it.

**Comment by Eric89GXL** [comment]
fix `mne/io/bti/bti.py:25:1: F401 'RAW_INFO_FIELDS' imported but unused` and this should be good to go

**Comment by dengemann** [comment]
More to come here, looking into the open / with function. But apparently a bug in numpy is being a blocker here. https://github.com/numpy/numpy/issues/2230 np.fromfile does not accept (anymore) StringIO objects.

**Comment by dengemann** [comment]
@Eric89GXL @agramfort last commits introduce support for reading from BytesIo.

File: mne/io/bti/bti.py [hunk 3]
```diff
@@ -43,6 +43,28 @@
...
+
+
+def bti_open(fname, *args, **kwargs):
+    """Handle bytes io"""
+    if isinstance(fname, six.string_types):
+        return open(fname, *args, **kwargs)
+    elif isinstance(fname, six.BytesIO):
+        return bytes_io_mock_context(fname)
+    else:
+        raise RuntimeError('Cannot mock this.')
```
Comment (dengemann) [inline mne/io/bti/bti.py]: To get the BytesIo pass the with and open blocks we need to hack a bit as you see.

File: mne/io/bti/bti.py [hunk 4]
```diff
@@ -952,13 +982,18 @@ def _read_data(info, start=None, stop=None):
...
         cnt = (stop - start) * info['total_chans']
         shape = [stop - start, info['total_chans']]
-        data = np.fromfile(fid, dtype=info['dtype'],
-                           count=cnt).astype('f4').reshape(shape)
+
+        if isinstance(fid, six.BytesIO):
+            data = np.fromstring(fid.getvalue(),
+                                 dtype=info['dtype'], count=cnt)
+        else:
+            data = np.fromfile(fid, dtype=info['dtype'], count=cnt)
```
Comment (dengemann) [inline mne/io/bti/bti.py]: fromfile won't work with a mocked fid, so we read strings instead.

File: mne/io/bti/bti.py [hunk 5]
```diff
@@ -1020,200 +1085,190 @@ def __init__(self, pdf_fname, config_fname='config',
...
-            if not op.isfile(head_shape_fname):
-                raise ValueError('Could not find the head_shape file "%s". '
-                                 'You should check whether you are in the '
-                                 'right directory or pass the full file name.'
-                                 % orig_name)
+        if not op.isfile(head_shape_fname):
+            raise ValueError('Could not find the head_shape file "%s". '
+                             'You should check whether you are in the '
+                             'right directory or pass the full file name.'
+                             % orig_name)
```
Comment (dengemann) [inline mne/io/bti/bti.py]: Above we skip checking filenames if it's a bytes io.

File: mne/io/bti/bti.py [hunk 6]
```diff
@@ -43,6 +43,28 @@
...
+class bytes_io_mock_context():
+
+    def __init__(self, target):
+        self.target = target
+
+    def __enter__(self):
+        return self.target
+
+    def __exit__(self, type, value, tb):
+        pass
```
Comment (dengemann) [inline mne/io/bti/bti.py]: maybe this one should be private.

**Comment by dengemann** [comment]
bummer, windows errors:

```
======================================================================
ERROR: Test bti bytes-io API
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Python27-conda64\lib\site-packages\nose\case.py", line 197, in runTest
    self.test(*self.arg)
  File "C:\projects\mne-python\mne\io\bti\tests\test_bti.py", line 217, in test_bytes_io
    raw2 = read_raw_bti(pdf, config, hs, convert=True)
  File "<string>", line 2, in read_raw_bti
  File "C:\projects\mne-python\mne\utils.py", line 540, in verbose
    return function(*args, **kwargs)
  File "C:\projects\mne-python\mne\io\bti\bti.py", line 1329, in read_raw_bti
    verbose=verbose)
  File "<string>", line 2, in __init__
  File "C:\projects\mne-python\mne\utils.py", line 540, in verbose
    return function(*args, **kwargs)
  File "C:\projects\mne-python\mne\io\bti\bti.py", line 1051, in __init__
    eog_ch=eog_ch)
  File "C:\projects\mne-python\mne\io\bti\bti.py", line 1102, in _get_bti_info
    pdf_fname, config_fname, sort_by_ch_name=sort_by_ch_name)
  File "C:\projects\mne-python\mne\io\bti\bti.py", line 838, in _read_bti_header
    info = {'version': read_int16(fid),
  File "C:\projects\mne-python\mne\io\bti\read.py", line 73, in read_int16
    return _unpack_simple(fid, '>' + ('H' * count),  count)
  File "C:\projects\mne-python\mne\io\bti\read.py", line 26, in _unpack_simple
    data = list(struct.unpack(fmt, string))
error: unpack requires a string argument of length 2
-------------------- >> begin captured stdout << ---------------------

```

Not sure when/how to debug this, no windows computer in reach for me.

**Comment by Eric89GXL** [comment]
I can look tomorrow if nobody beats me to it. In the meantime you could try
refactoring to use numpy fromstring instead

**Comment by dengemann** [comment]
> In the meantime you could try
refactoring to use numpy fromstring instead

it's already done eric.

**Comment by dengemann** [comment]
> refactoring to use numpy fromstring instead

ok I start seeing what you have in mind. ...

**Comment by agramfort** [comment]
I don't understand much of the BTi code. I cannot really help here :(

File: mne/io/bti/bti.py (hunk 6 above)
Comment (Eric89GXL) [inline mne/io/bti/bti.py]: yeah private is better I think

**Comment by Eric89GXL** [comment]
@dengemann the error was actually on Linux, too (any Py3k build should have it):

https://travis-ci.org/mne-tools/mne-python/jobs/82299244

You just needed `with open(..., 'rb')` instead of `with open(...)` in a couple of places, I pushed a commit that should fix it.

If you get bored at some point, it's probably still worth changing the read functions to use `np.fromstring` at some point, it should simplify/unify reading, since it's what we do most other places in the package.

**Comment by Eric89GXL** [comment]
Looks like we use `struct` in ~7 places, might be worth getting rid of them. Maybe I'll do it Monday if I have a free few minutes.

**Comment by dengemann** [comment]
@Eric89GXL cool, perfect.

**Comment by agramfort** [comment]
ready to go?

**Comment by dengemann** [comment]
we need to make the mocker class private and eric suggested to get rid of struct unpack. I will look into it nowish.

**Comment by Eric89GXL** [comment]
I can do the struct unpacking stuff tomorrow package-wide since it's unrelated anyway. But if you have time today feel free to go for it.

**Comment by dengemann** [comment]
Almost done.

**Comment by dengemann** [comment]
> I can do the struct unpacking stuff tomorrow package-wide since it's unrelated anyway. But if you have time today feel free to go for it.

done

**Comment by Eric89GXL** [comment]
There are still instances of `struct` in `mne/io/tag.py`, `tree.py`, and `mne/realtime/client.py`

**Comment by dengemann** [comment]
> here are still instances of struct in mne/io/tag.py, tree.py, and mne/realtime/client.py

that's for another PR then, no? This is bti.

**Comment by Eric89GXL** [comment]
>> I can do the struct unpacking stuff tomorrow **package-wide** since it's unrelated anyway. But if you have time today feel free to go for it.

> done

Not quite, then :)

**Comment by dengemann** [comment]
@Eric89GXL did not see "package-wide".

**Comment by dengemann** [comment]
Travis is green, I'm happy, you can merge if you're happy.

**Comment by Eric89GXL** [comment]
Awesome, thanks @dengemann

**Pull request merged**